pkg/cloudprovider/vsphereparavirtual/vmservice/vmservice_test.go

```diff
@@ -41,669 +40,595 @@ import (
+			assert.NoError(t, err)
+			assert.Equal(t, (*vmServiceObj).Spec, testCase.expectedSpec)
+
+			testCase.testK8sService.Spec.LoadBalancerIP = ""
+			testCase.testK8sService.Spec.LoadBalancerSourceRanges = []string{}
+			err = vms.Delete(context.Background(), testCase.testK8sService, testClustername)
+			assert.NoError(t, err)
 		})
+	}
+}
+
+func TestCreateVMService_ExternalTrafficPolicyTypeLocal(t *testing.T) {
+	testK8sService, vms, _ := initTest()
+	testK8sService.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeLocal
+	testK8sService.Spec.HealthCheckNodePort = 30012
+	vmServiceObj, err := vms.Create(context.Background(), testK8sService, testClustername)
+	assert.NoError(t, err)
+
+	v, ok := vmServiceObj.Annotations[AnnotationServiceExternalTrafficPolicyKey]
+	assert.Equal(t, ok, true)
+	assert.Equal(t, v, string(v1.ServiceExternalTrafficPolicyTypeLocal))
+
+	hcPort, ok := vmServiceObj.Annotations[AnnotationServiceHealthCheckNodePortKey]
+	assert.Equal(t, ok, true)
+	assert.Equal(t, hcPort, strconv.Itoa(30012))
+
+	testK8sService.Spec.ExternalTrafficPolicy = ""
+	err = vms.Delete(context.Background(), testK8sService, testClustername)
+	assert.NoError(t, err)
+}
+
+func TestCreateVMService_ExternalTrafficPolicyTypeCluster(t *testing.T) {
+	testK8sService, vms, _ := initTest()
+	testK8sService.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeCluster
+	vmServiceObj, err := vms.Create(context.Background(), testK8sService, testClustername)
+	assert.NoError(t, err)
+
+	_, ok := vmServiceObj.Annotations[AnnotationServiceExternalTrafficPolicyKey]
+	assert.NotEqual(t, ok, true)
+
+	_, ok = vmServiceObj.Annotations[AnnotationServiceHealthCheckNodePortKey]
+	assert.NotEqual(t, ok, true)
+
+	testK8sService.Spec.ExternalTrafficPolicy = ""
+	err = vms.Delete(context.Background(), testK8sService, testClustername)
+	assert.NoError(t, err)
+}
+
+func TestVMService_NotExist(t *testing.T) {
+	_, vms, _ := initTest()
+	k8sService := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      testK8sServiceName,
+			Namespace: testK8sServiceNameSpace,
+		},
+	}
+	_, err := vms.CreateOrUpdate(context.Background(), k8sService, testClustername)
+	assert.Error(t, err)
+	assert.Equal(t, err, ErrVMServiceIPNotFound)
+}
+
+func TestVMService_EmptyClusterName(t *testing.T) {
+	testK8sService, vms, _ := initTest()
+	_, err := vms.CreateOrUpdate(context.Background(), testK8sService, "")
+	assert.Error(t, err)
+}
+
+func TestVMService_ReturnErr(t *testing.T) {
+	testK8sService, vms, fcw := initTest()
+	// Redefine Get in the client to return an error
+	fcw.GetFunc = func(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+		return fmt.Errorf("failed to get VirtualMachineService")
+	}
+	_, err := vms.CreateOrUpdate(context.Background(), testK8sService, testClustername)
+	assert.Error(t, err)
+}
+
+func TestVMService_CreationFails(t *testing.T) {
+	testK8sService, vms, fcw := initTest()
```

Comment:
I think the original test cases is not well defined, this should be split into two cases

Reply:
can you explain a little more?
should we remove 
```
// Redefine Get in the client to return an error
	fcw.GetFunc = func(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
		return apierrors.NewNotFound(v1alpha1.Resource("virtualmachineservice"), testClustername)
	}
```?

Reply:
I think it's two cases  that one failed at Get, one failed at Create, you can create two cases testing this func